### PR TITLE
Gcs upload copy

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Improve gcs-upload performace with busy targets

gcs-upload uses `rsync` to upload folders. When the target contains
many thousands of files, it can take a very long time.

Change the default implementation to use `cp -r` which only copies
new files. Recursive copy supports not replacing existing files,
so adding an option to the task for that.

The option to delete extra remote files is still there, and still
false by default. Delete extra remote files is not compatible with
"not replace remote files" so adding a note about that.

Signed-off-by: Andrea Frittoli <andrea.frittoli@gmail.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Follows the [authoring recommendations](https://github.com/tektoncd/catalog/blob/main/recommendations.md)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality of task changed or new task added)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Complies with [Catalog Organization TEP][TEP], see [example]. **Note** [An issue has been filed to automate this validation][validation]
  - [x] File path follows  `<kind>/<name>/<version>/name.yaml`
  - [x] Has `README.md` at `<kind>/<name>/<version>/README.md`
  - [x] Has mandatory `metadata.labels` - `app.kubernetes.io/version` the same as the `<version>` of the resource
  - [x] Has mandatory `metadata.annotations` `tekton.dev/pipelines.minVersion`
  - [x] mandatory `spec.description` follows the convention

          ```

          spec:
            description: >-
              one line summary of the resource

              Paragraph(s) to describe the resource.
          ```

_See [the contribution guide](https://github.com/tektoncd/catalog/blob/master/CONTRIBUTING.md)
for more details._

---

[TEP]: https://github.com/tektoncd/community/blob/master/teps/0003-tekton-catalog-organization.md
[example]: https://github.com/tektoncd/catalog/tree/master/task/git-clone/0.1
[validation]:  https://github.com/tektoncd/catalog/issues/413

### CODE REVIEW

**Comment by afrittoli** [comment]
> Catlin Output
> ```
> FILE: task/gcs-upload/0.3/gcs-upload.yaml
> ERROR: Resource path is invalid; expected path: task/gcs-upload/0.2/gcs-upload.yaml
> ```

I don't understand this catlin output - the path seems ok to me?

File: task/gcs-upload/0.3/gcs-upload.yaml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+              echo "\"deleteExtraFiles\"" can be used only when \"replaceExistingFiles\"" is true"
+              exit 1
+            fi
+          # When deleteExtraFiles is requested, use rsync
+          gsutil -m rsync -d -r "${SOURCE}" "${LOCATION}"
+        else
+          CP_PARAMS=""
+          if [[ "$REPLACE_EXISTING_FILES" == "false" ]]; then
+              CP_PARAMS="-n"
+          fi
```
Comment (vinamra28) [inline task/gcs-upload/0.3/gcs-upload.yaml:100]: the following lines are duplication of [this](https://github.com/tektoncd/catalog/pull/964/files#diff-e07dc48d127c7bc7ae6fa765137f8cff6d68da71ba9b2169afad123ddf9cfaf9R82-R85) I guess

Comment (afrittoli) [reply task/gcs-upload/0.3/gcs-upload.yaml:100]: oh, nice catch, I moved it up and forgot to delete it

Comment (afrittoli) [reply task/gcs-upload/0.3/gcs-upload.yaml]: Fixed now

**Comment by vinamra28** [comment]
/lgtm
Thanks 😄

**Pull request merged**